llvm/lib/Target/DirectX/DXILOpLowering.cpp

```diff
@@ -461,6 +479,27 @@ class OpLowerer {
     });
   }
 
+  [[nodiscard]] bool lowerSplitDouble(Function &F) {
+    IRBuilder<> &IRB = OpBuilder.getIRB();
+    return replaceFunction(F, [&](CallInst *CI) -> Error {
+      IRB.SetInsertPoint(CI);
+
+      Value *Arg0 = CI->getArgOperand(0);
+
+      Type *NewRetTy = OpBuilder.getResSplitDoubleType(M.getContext());
+
+      std::array<Value *, 1> Args{Arg0};
+      Expected<CallInst *> OpCall = OpBuilder.tryCreateOp(
```

Comment:
could you explain how this line and DXIL.td  work together? You are `specifying `OpCode::SplitDouble` in the createOp and on line 531 you are mapping the `Intrinsic::dx_splitdouble:` intrinsic  to this lowering behavior. Does that mean DXIL.td isn't used at all?

Reply:
`DXIL.td`, is defining the OpCode call signature, like it's arguments, return type, shader stages where it is available, etc.
First, the `OpCode` is generated from `DXIL.td`, so the Op codes defined in there will be made available through the `OpCode` class. Finally, when we call `tryCreateOp` it will check the values specified within `DXIL.td` in order to make sure the call is valid.

Reply:
so what im wondering is it possible to just something that looks like:
```
case Intrinsic::dx_splitdouble:
replaceSplitDoubleCallUsages(...);
replaceFunctionWithOp(F, OpCode); 
```
instead of writing our own intrinsic to op code lowering as a one off.

Reply:
We could potentially do this in a semi-generic way by having a `replaceFunctionWithNamedStructOp(F, OpCode, NewRetTy)`. Conceptually, we could even do that automatically within `replaceFunctionWithOp` with a little bit more tablegen. For these layout-compatible-but-named-struct operations I would expect them to all be the same.

Reply:
I am able to create `replaceFunctionWithNamedStructOp(F, OpCode, NewRetTy, ReplaceUsesLambda)`. The ReplaceUsesLambda will be used to deal with the replacement of uses logic, which is something that we need to deal with in a function basis. I've updated the code to and example with of such function.